### ISSUE TITLE
What's the best way to pass a value from one pipeline stage to the next

### ISSUE THREAD

**Issue by retronym** [scala/scala-jenkins-infra#237]
Historically, we used the BuildFlow plugin to orchestrate our build pipelines for pull request validation and for release builds.

We're in the process of migrating that to a Jenkins 2.0 pipeline.

This pipeline is representative of our current solution.

```groovy
import java.nio.file.*

node {
    branch = '2.12.x'
    paramList = [
        [$class: 'StringParameterValue', name: 'repo_user', value: "${repo_user}"],
    ]
    testParamList = [[$class: 'StringParameterValue', name: 'dummy', value: 'dummyvalue']]
    stage('test stage') {
        built = build(job: "test-job", parameters: [[$class: 'StringParameterValue', name: 'value', value: 'hola']])
        copyArtifacts(projectName: 'test-job', selector: specific("${built.number}"), filter: '**/artifact.txt', flatten: true);
        readValue = readFile(Paths.get("${env.WORKSPACE}/artifact.txt")).trim()
        println("Read [$readValue] from copied artifact")
        testParamList = testParamList + [$class: 'StringParameterValue', name: "value", value: ("¡" + readValue + "!")]
    }
    stage('test stage 2') {
        parallel (
            sub1: { node {
                build(job: "test-job", parameters: testParamList)
            }},
            sub2: { node {
                build(job: "test-job", parameters: testParamList)
            }}
        )
    }
}

def readFile(Path path) {
    return Files.readAllLines(path).stream().collect(java.util.stream.Collectors.joining("\n"))
}
```

The build we trigger is run on a different machine from the pipeline.

In the real use case, the first builds artifacts and publishes them to a repository, and writes the (computed) version number to the artifact file. The pipeline reads the artifact, and passes that version number as a parameter to the subsequent builds.

Because we're doing some low-level work to thread `built.number` into the `copyArtifacts` step, we're using the Groovy Scripted DSL. Will this limit us at all when it comes to using the new [Blue Ocean](https://jenkins.io/projects/blueocean/) UX? 

What is the standard way of doing this in Jenkins. [Stash/Unstash](https://jenkins.io/doc/pipeline/steps/workflow-basic-steps/#code-stash-code-stash-some-files-to-be-used-later-in-the-build)? [`copyArtifacts`](https://jenkins.io/doc/pipeline/steps/copyartifact/) somehow attached to the downstream project configured with the `upstream` selector? I'm pretty baffled after a day of looking at this.

**Comment by retronym** [comment]
No clear answer in https://stackoverflow.com/questions/39730193/pass-artifact-or-string-to-upstream-job-in-jenkins-pipeline either, although it does have a suggestion to use `buildVariables`.

**Comment by retronym** [comment]
A bit more info on `buildVariables`. https://issues.jenkins-ci.org/browse/JENKINS-36528

**Comment by retronym** [comment]
I can't seem to set a dynamic entry into the downstream job's environment. If I could, it would be made available to the orchestrating pipeline via `def built = build ...; built.buildVariables`.

Here's what I tried, as, a System Groovy Script build step in the downstream freestyle job:

```groovy
import java.nio.file.*

def readFile(Path path) {
    return Files.readAllLines(path).stream().collect(java.util.stream.Collectors.joining("\n"))
}

value = readFile(Paths.get(build.workspace.toString(), "artifact.txt"))
println(value)
env = build.getEnvironment(listener)
env.put("V1", value)
println(env.get("V1")) // prints the value
println(build.getEnvironment(listener).get("V1")) // null
```

**Comment by retronym** [comment]
Another thought: maybe using the `build` step is a pipeline smell, and we should just inline the definitions of the sub-builds (they are all simple, just calls to builds scripts in the repo) into the pipeline definition.

[Best Practice `#10`](https://www.cloudbees.com/blog/top-10-best-practices-jenkins-pipeline-plugin) suggests that `stash` is the preferred way to shuffle state between parts of the pipeline, and I guess that is only a possibility if everything is expressed in the pipeline.

**Comment by retronym** [comment]
Here's what I'm experimenting with to use "inlined" configurations and stash/unstash to share the file. This seems to give us enough power to configure multi-node parallelism, without the baggage of physical Jobs for the leaf tasks.

I need to [learn a bit more](https://jenkins.io/doc/book/pipeline/shared-libraries/) about the DSL configuration to DRY up the config (e.g. we're repeating the git checkout incantation)

```groovy
#!groovy
pipeline {
    agent none
    options {
      buildDiscarder(logRotator(numToKeepStr:'10'))
      timeout(time: 120, unit: 'MINUTES')
    }
    stages {
        stage('Publish') {
            agent {
               label ""
            }
            tools {
                jdk '1.8.0' 
            }
            steps {
                checkout scm: [$class: 'GitSCM', 
      userRemoteConfigs: [[url: 'git@github.com:scala/scala.git']], 
      branches: [[name: '2.13.x']]], changelog: false, poll: false
                ansiColor('xterm') {
                    sh "/usr/local/bin/sbt -mem 2G setupPublishCore publishLocal generateBuildCharacterPropertiesFile"
                }
                stash includes: 'buildcharacter.properties', name: 'version_props'
            }
        }
    
        stage('Test') {
            parallel {
                stage('Junit') {
                    agent {
                       label ""
                    }
                    tools {
                        jdk '1.8.0' 
                    }
                    steps {
                        checkout scm: [$class: 'GitSCM', 
              userRemoteConfigs: [[url: 'git@github.com:scala/scala.git']], 
              branches: [[name: '2.13.x']]], changelog: false, poll: false
        
                        unstash name: 'version_props'
                        ansiColor('xterm') {
                            sh "/usr/local/bin/sbt -mem 2G -Dstarr.version=\$(grep maven.version.number buildcharacter.properties | cut -d = -f 2) junit/test"
                        }
                    }
                }
                stage('partest') {
                    agent {
                       label ""
                    }
                    tools {
                        jdk '1.8.0' 
                    }
                    steps {
                        checkout scm: [$class: 'GitSCM', 
              userRemoteConfigs: [[url: 'git@github.com:scala/scala.git']], 
              branches: [[name: '2.13.x']]], changelog: false, poll: false
        
                        unstash name: 'version_props'
                        ansiColor('xterm') {
                            sh "/usr/local/bin/sbt -mem 2G -Dstarr.version=\$(grep maven.version.number buildcharacter.properties | cut -d = -f 2) 'partest --neg --terse' "
                        }
                    }
                }
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/65551/35918394-5f7fc47c-0c5d-11e8-9c49-93e5488a96c2.png)

![image](https://user-images.githubusercontent.com/65551/35918567-023445da-0c5e-11e8-9f64-d9fd09398dbc.png)

**Comment by adriaanm** [comment]
Great! Let's move this into a Jenkinsfile that's tracked in scala/scala and check it out in the *-main jobs. All the other ones can go. Right?

**Comment by michaelneale** [comment]
downstream build steps are very popular, and I wouldn't say a "smell" but they aren't supported as well as they should be. Pipeline was built mostly to be self containing - ie the pipeline itself runs end to end, any downstream runs that are fired off are very very loosely coupled. To avoid inlining of duplicated stuff, a pipeline lib may often be used so the common globs of steps are shared in common, but each pipeline run stands alone.

**Comment by retronym** [comment]
Thanks, Michael. The self-contained model looks to be working well enough for us, and lets us remove a bunch of boilerplate downstream jobs.

Here's the latest incarnation of the PR validation pipeline: https://gist.github.com/62c06f63549c8da4f71d2e196842b490

It's pretty basic stuff, save for the complication of needing to bootstrap the compiler and setting ourselves up to run test jobs in parallel.

While I've got your attention, is it a known problem to have the [stage view](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main-experimental/workflow-stage/) (screenshot below) showing all success when the build is still running? The [pipeline steps](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main-experimental/16/flowGraphTable/) view shows the status correctly.

![image](https://user-images.githubusercontent.com/65551/35955545-b43b8d1a-0cdb-11e8-80b3-98a800f8a631.png)

I've come across a few won't fix tickets about stage view pointing to Blue Ocean so I tried that out locally. That does a much better job of rendering branching pipelines. It seems pretty heavy on CPU in Chrome in the build progress view (even with logs collapsed), which hopefully is a teething problem; I look forward to seeing it mature.

**Comment by retronym** [comment]
> Great! Let's move this into a Jenkinsfile that's tracked in scala/scala and check it out in the *-main jobs. All the other ones can go. Right?

I _think_ that's an improvement, and it is going further in the direction we've already taken by having dumb build jobs that just call `./scripts/validate/{publish,test}` etc. If we push more into a Jenkinsfile, it will give us the flexibility to split up the `test` script into smaller chunks (`testJunit / testBinaryComat` etc) which could be run in parallel across the build farm, but we wouldn't need to bake knowledge of the name of those chunks into the Jenkins definition. So that all sounds good to me.

My only concern is retaining the control to make certain types of cross cutting changes/fixes/improvements to the build without needing to change the jenkinsfile on each branch. Things like adding a build parameter to control _JAVA_OPTIONS, or changing the way we set the active JDK version on the slaves.

**Comment by retronym** [comment]
I think a good middle ground is use [this technique](https://devops.stackexchange.com/a/3090) to dynamically create parallel stages for all scripts in:

```
scripts/jobs/validate/test,scripts/jobs/validate/test/*,scripts/jobs/integrate/ide
```

We can then easily experiment with changing the granularity of the test jobs by splitting up `validate/test` into `validate/test/{junit,mima,run-jvm,pos-neg,misc}`. The jenkins job will be backwards compatible with older commits (importantly, in-flight PRs).

**Comment by michaelneale** [comment]
@retronym I think that is known - normally there is a little visual cue that is is actively running on stage view
